### PR TITLE
Additional properties - UI and Backend

### PR DESCRIPTION
**Pull request by kavishkafernando**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Details/Overview/AdditionalProperty.jsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Details/Overview/AdditionalProperty.jsx]: Resolved

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 2]
```diff
@@ -2753,6 +2760,60 @@ private void deleteSubscriptionPolicies(Connection connection, String apiID) thr
...
 
+    /**
+     * Add additional properties to the database
+     *
+     * @param connection SQL Connection
+     * @param apiID      ApiId of the API
+     * @param additionalProperty      List<AdditionalProperties>
+     * @throws SQLException If failed to add additional properties
+     */
+    private void addAdditionalProperties(Connection connection, List<AdditionalProperties> additionalProperty,
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: Resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/utils/MappingUtil.java [hunk 3]
```diff
@@ -155,6 +157,14 @@ public static APIDTO toAPIDto(API api) throws APIManagementException {
             apiOperationsDTO.setScopes(uriTemplate.getScopes());
             apidto.addOperationsItem(apiOperationsDTO);
         }
+
+        if (api.getAdditionalProperties() != null) {
+            List<API_additionalPropertiesDTO> additionalPropertiesDTOList = new ArrayList<>();
+            for (AdditionalProperties propertiesDTO : api.getAdditionalProperties()){
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/utils/MappingUtil.java]: Resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/utils/MappingUtil.java [hunk 4]
```diff
@@ -270,6 +280,15 @@ public static APIDTO toAPIDto(API api) throws APIManagementException {
                 scopes(apidto.getScopes()).
                 securityScheme(mapSecuritySchemeListToInt(apidto.getSecurityScheme()));
 
+        //additional properties
+        if (apidto.getAdditionalProperties() != null) {
+            List<AdditionalProperties> additionalProperties = new ArrayList<>();
+            //get additional properties for each list element in API_additionalPropertiesDTO
+            for (API_additionalPropertiesDTO propertiesDTO : apidto.getAdditionalProperties()){
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/utils/MappingUtil.java]: Resolved

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/models/AdditionalProperties.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
+ *
+ *   Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/models/AdditionalProperties.java]: Resolved

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java [hunk 6]
```diff
@@ -2753,6 +2760,60 @@ private void deleteSubscriptionPolicies(Connection connection, String apiID) thr
...
+                                         String apiID) throws SQLException {
+        final String query =
+                "INSERT INTO AM_API_ADDITIONAL_PROPERTIES(UUID, PROPERTY_KEY, PROPERTY_VALUE) VALUES (?, ?, ?)";
+        try (PreparedStatement statement = connection.prepareStatement(query)) {
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/dao/impl/ApiDAOImpl.java]: resolved

**Comment by pubudu538** [comment]
Let's add an integration test for this feature in product-apim. https://github.com/wso2/product-apim/tree/master/modules/integration-tests

**Pull request merged**